executor/slow_query.go

```diff
@@ -186,35 +183,136 @@ func (sc *slowLogChecker) isTimeValid(t types.Time) bool {
+	} else {
+		resByte = make([]byte, len(lineByte))
+	}
+	// Use copy here to avoid shallow copy problem.
+	copy(resByte, lineByte)
+	if err != nil {
+		return resByte, err
+	}
+
+	var tempLine []byte
+	for isPrefix {
+		tempLine, isPrefix, err = reader.ReadLine()
+		resByte = append(resByte, tempLine...)
+		// Use the max value of max_allowed_packet to check the single line length.
+		if len(resByte) > int(variable.MaxOfMaxAllowedPacket) {
+			return resByte, errors.Errorf("single line length exceeds limit: %v", variable.MaxOfMaxAllowedPacket)
 		}
-		e.fileLine++
-		lineByte, err := getOneLine(reader)
 		if err != nil {
-			if err == io.EOF {
-				e.fileIdx++
-				e.fileLine = 0
-				if e.fileIdx >= len(e.files) {
-					return rows, nil
+			return resByte, err
+		}
+	}
+	return resByte, err
+}
+
+type offset struct {
+	offset int
+	length int
+}
+
+func (e *slowQueryRetriever) getBatchLog(reader *bufio.Reader, offset *offset, num int) ([]string, error) {
+	var line string
+	log := make([]string, 0, num)
+	var err error
+	for i := 0; i < num; i++ {
+		for {
+			e.fileLine++
+			lineByte, err := getOneLine(reader)
+			if err != nil {
+				if err == io.EOF {
+					e.fileIdx++
+					e.fileLine = 0
+					if e.fileIdx >= len(e.files) {
+						return log, nil
+					}
+					offset.length = len(log)
+					reader.Reset(e.files[e.fileIdx].file)
+					continue
 				}
-				reader.Reset(e.files[e.fileIdx].file)
-				continue
+				return log, err
+			}
+			line = string(hack.String(lineByte))
+			log = append(log, line)
+			if strings.HasSuffix(line, variable.SlowLogSQLSuffixStr) {
+				if strings.HasPrefix(line, "use") {
+					continue
+				}
+				break
+			}
+		}
+	}
+	return log, err
+}
+
+func (e *slowQueryRetriever) parseSlowLog(ctx context.Context, sctx sessionctx.Context, reader *bufio.Reader, logNum int) {
+	var wg sync.WaitGroup
+	offset := offset{offset: 0, length: 0}
+	// To limit the num of go routine
+	ch := make(chan int, sctx.GetSessionVars().Concurrency.DistSQLScanConcurrency())
+	defer close(ch)
+	for {
+		log, err := e.getBatchLog(reader, &offset, logNum)
+		if err != nil {
+			e.parsedSlowLogCh <- parsedSlowLog{nil, err}
+			break
+		}
+		start := offset
+		wg.Add(1)
+		ch <- 1
+		go func() {
+			defer wg.Done()
+			result, err := e.parsedLog(sctx, log, &start)
+			if err != nil {
+				e.parsedSlowLogCh <- parsedSlowLog{nil, err}
+			} else {
+				e.parsedSlowLogCh <- parsedSlowLog{result, err}
 			}
-			return rows, err
+			<-ch
+		}()
+		// Read the next file, offset = 0
+		if e.fileIdx >= len(e.files) {
+			break
+		}
+		offset.offset = e.fileLine
+		offset.length = 0
+		select {
+		case <-ctx.Done():
+			break
+		default:
+		}
+	}
+	wg.Wait()
+}
+
+func (e *slowQueryRetriever) parsedLog(ctx sessionctx.Context, log []string, offset *offset) (data [][]types.Datum, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("%s", r)
+		}
+	}()
+	failpoint.Inject("errorMockPanic", func(val failpoint.Value) {
```

Comment:
```suggestion
	failpoint.Inject("errorMockParseSlowLogPanic", func(val failpoint.Value) {
```

Better to use a unique failpoint name.